### PR TITLE
fix(Postgres Node): Re-use connection pool across executions

### PR DESCRIPTION
**Pull request by despairblue**
## Summary

This add a connection pool manager similar to the ssh [SSHClientsManager](https://github.com/n8n-io/n8n/blob/fed2b9004a7e70206fcb7cbce6effd9cf642f493/packages/core/src/SSHClientsManager.ts#L7) for re-using ssh connections.

It's a singleton that lives in the nodes-base package. Adi and I initially tried to make this available through the context (node execution functions) and also have this available to the credentials and have them manage the connections, but this led to changes to the interfaces in the workflow package that would have culminated in a single huge refactor in this PR. Instead we chose this simpler approach that can later be integrated into the context and be made available to the credentials instead of the nodes by further refactors.

Next step is to use the manager for more nodes, e.g. MSSQL and MySQL. This will also fix an issue that at least one cloud customer has with connections to MSSQL being blocked because we connect and disconnect too often when they execute the MSSQL node in a tight loop.

## Related Linear tickets, Github issues, and Community forum posts

todo: add help ticket link

## Review / Merge checklist

- [x] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) 
- [x] ~[Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.~
- [x] Tests included. 
- [x] ~PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)~

### CODE REVIEW

File: packages/nodes-base/utils/connection-pool-manager.ts [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+			string,
+			{ pool: unknown; cleanUpHandler: (pool: unknown) => Promise<void>; lastUsed: number }
+		> = new Map(),
+	) {
+		// Close all SSH connections when the process exits
+		process.on('exit', () => this.onShutdown());
+
+		// Regularly close stale SSH connections
+		setInterval(() => this.cleanupStaleConnections(), cleanUpInterval);
+	}
```
Comment (netroy) [inline packages/nodes-base/utils/connection-pool-manager.ts:47]: having this in the constructor is a bit confusing. can we move this bit out?
```suggestion
	private map: Map<
			string,
			{ pool: unknown; cleanUpHandler: (pool: unknown) => Promise<void>; lastUsed: number }
		> = new Map()

	/**
	 * Private constructor that initializes the connection pool manager.
	 * Sets up cleanup handlers for process exit and stale connections.
	 */
	private constructor() {
		// Close all SSH connections when the process exits
		process.on('exit', () => this.onShutdown());

		// Regularly close stale SSH connections
		setInterval(() => this.cleanupStaleConnections(), cleanUpInterval);
	}
```

also, we could extract the `{ pool: unknown; cleanUpHandler: (pool: unknown) => Promise<void>; lastUsed: number }` bit into a type, and add jsdocs for each property (for future readability).

File: packages/nodes-base/utils/connection-pool-manager.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+	/**
+	 * Gets or creates a connection pool for the given options.
+	 * Updates the last used timestamp for existing connections.
+	 */
+	async getConnection<T>(
+		options: Options,
+		fallBackHandler: () => Promise<T>,
+		cleanUpHandler: (pool: T) => Promise<void>,
```
Comment (netroy) [inline packages/nodes-base/utils/connection-pool-manager.ts:73]: should we create a new type `GetConnectionOptions` that extends `Options`, but had `fallBackHandler` and `cleanUpHandler` as fields?
That would make the call to `getConnection` in the nodes code a lot more readable.

**Review by netroy — CHANGES_REQUESTED** [review]
besides some minor comments, LGTM 🎉.

**Review by netroy — APPROVED** [review]

**Comment by janober** [comment]
Got released with [`n8n@1.74.1`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.74.1)

**Comment by janober** [comment]
Got released with [`n8n@1.74.1`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.74.1)

**Comment by janober** [comment]
Got released with [`n8n@1.74.0`](https://github.com/n8n-io/n8n/releases/tag/n8n%401.74.0)

**Comment by akhudolev** [comment]
Hi! It looks like this change caused issues with postgres connections with ssh tunnel
https://github.com/n8n-io/n8n/issues/12807

**Pull request merged**